gh-128555: Add 'context' keyword parameter to Thread. (#128209)
python/cpython: Doc/library/threading.rst

   If ``None`` (the default), the daemonic property is inherited from the
   current thread.

   *context* is the `contextvars.Context` value to use while running the thread.

>    The default is to inherit the context of the caller of :meth:`~Thread.start`.

Reviewer: I think passing `None` should by default should mean that the copy of the current context is being used. if empty context is required, users can just pass empty context `Context()`. This would make it behave similar to how it works in asyncio. 

https://docs.python.org/3.13/library/asyncio-task.html#asyncio.create_task

Author: Seems good to me to follow what asyncio does here, changed.

   If ``None`` (the default), the daemonic property is inherited from the
   current thread.

   *context* is the `contextvars.Context` value to use while running the thread.

>    The default is to inherit the context of the caller of :meth:`~Thread.start`.

Reviewer 2: clarify this: "to inherit a copy of the context of"

   .. versionchanged:: 3.10
      Use the *target* name if *name* argument is omitted.

   .. versionchanged:: 3.14

>       Added the *context* parameter.  Previously threads always ran with an empty

Reviewer 2: I think I'd make this more direct about the behavior change, perhaps something like:

> "Threads now inherit the context of the caller of Thread.start instead of starting with an empty context. The context parameter was added. Pass a new contextvars.Context() if your thread requires an empty context."

This default change is a behavior change. Expect that to trip someones existing code up. I don't have the context _(pun intended)_ as to how disruptive that could be. Per the [PEP-387 breaking change policy](https://peps.python.org/pep-0387/) we'd want to wait at least two releases. Which isn't so satisfying given the reasons you want this.

But a compromise could be considered (unsure if this is really _wise_) if needed: Change the default sooner than the deprecation when running in an experimental free-threading cpython build?

Author: I asked Sam what he thought about making this behavior enabled by default on free-threaded builds and he's okay with the idea.  I've updated this PR to take that approach.  I think we can wait a couple of releases to decide what to do about the default build.  If it seems there is little breakage, we can change the default there too after a suitable period.

   .. versionchanged:: 3.10
      Use the *target* name if *name* argument is omitted.

   .. versionchanged:: 3.14
      Threads now inherit the context of the caller of :meth:`~Thread.start`

>       instead of starting with an empty context.  The ``context`` parameter

Reviewer: ```suggestion
      instead of starting with an empty context.  The *context* parameter
```

>       was added.  Pass a new :class:`~contextvars.Context()` if your thread
>       requires an empty context.

Reviewer: ```suggestion
      was added.  Pass a new :class:`~contextvars.Context()` if an empty context
      is required. 
```
